### PR TITLE
feat: Quote Wrapper Pattern

### PR DESCRIPTION
**Pull request by jmuzina**
## Done

- Implements Quote Wrapper pattern per [design](https://www.figma.com/design/Hll6ZRvHyTTGBZZSoRd8ei/Higher-order-components-(HOCs)?node-id=1974-5068&m=dev) and [spec](https://docs.google.com/document/d/1-TMWuLvxmOKxPb8gITCEna6RVg05eGHKsB0EdcJ1w6Y/edit?tab=t.0)
- Drive-bys:
  - Adds notes to all patterns docs pages about the patterns being a recent/in-flux addition

Fixes [WD-13881](https://warthogs.atlassian.net/browse/WD-13881)

## QA
- Review all quote wrapper examples
  - **Large**
    - [All elements](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large)
    - [Without signpost](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-no-signpost)
    - [Without citation](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-no-citation)
    - [Without header](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-no-header)
    - [Without signpost and without header](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-no-header-no-signpost)
    - [Without CTA or image](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-no-cta-image?theme=light)
    - [Without CTA, image, or citation](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-no-cta-image-or-citation?theme=light)
    - [Minimal](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-minimal)
  - **Medium**
    - [All elements](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/medium)
    - [Without signpost](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/medium-no-signpost)
    - [Without citation](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/medium-no-citation)
    - [Without header](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/medium-no-header)
  - **Small**
    - [All elements](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/small?theme=light)
    - [Without signpost](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/small-no-signpost)
    - [Without citation](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/small-no-citation)
    - [Without header](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/small-no-header)
- Review [combined quote wrapper example](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/combined?theme=light) and verify all above examples are rendered as expected.
- Review [Quote Wrapper docs](https://vanilla-framework-5275.demos.haus/docs/patterns/quote-wrapper)
- Review updated [4.18.0 release notes](https://vanilla-framework-5275.demos.haus/docs/whats-new)

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

- [x] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [ ] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical/vanilla-framework/releases/latest), following semver convention:
  - if CSS class names are not changed it can be bugfix release (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [x] Any changes to component class names (new patterns, variants, removed or added features) should be listed on the [what's new page](https://github.com/canonical/vanilla-framework/blob/main/releases.yml).

## Screenshots
Screenshot 2024-09-27 at 2 13 58 PM
Screenshot 2024-09-27 at 2 14 07 PM
Screenshot 2024-09-27 at 2 14 23 PM

[WD-13881]: https://warthogs.atlassian.net/browse/WD-13881?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [inline templates/_macros/quote-wrapper.jinja:81]: We need to decide to either merge and use borderless CTA (from #5278) or just pass in link text and href and render a simple link here, without caring about text wrapping

Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja]: Adjusted this to simply use a slot for passing in link content

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja]: Good call - adjusted

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja:16]: I'm hesitant to abstract this that far, in case design at some point wants us to support quote with text size that is not `<h2>` or `<h4>`.

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja:18]: I suppose it's possible that we could be quoting something said by multiple people or a larger organization, rather than a specific person. 

Maybe this could be `citation_source_name` for added clarity without discouraging people from citing people or larger organizations?

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja]: Updated to use unicode quotation mark entities

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja]: Good catch, not sure how that one got by the PR tests. Fixed!

File: templates/docs/examples/patterns/quote-wrapper/h2-no-citation.html
Comment (jmuzina) [reply templates/docs/examples/patterns/quote-wrapper/h2-no-citation.html]: Adjusted chevrons to use HTML entity

File: templates/docs/patterns/quote-wrapper/index.md
Comment (jmuzina) [reply templates/docs/patterns/quote-wrapper/index.md]: Updated to `Source name|title`

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja]: Updated to use `p-heading--` classes rather than directly manipulating the tag name

**Comment by jmuzina** [comment]
> There appears to be something going on here with the width of the `hr`/divider - at some screen sizes, it doesn't stretch all the way

@pastelcyborg Good catch, I think I've fixed that by moving the highlighted rule to outside of the grid, so it covers the full width of the whole pattern

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja]: Updated to `citation_source_name` and `citation_source_title`

**Comment by jmuzina** [comment]
Made two changes to align with parts of the design I missed:

- Citation source name and title are now split on medium
- There is now an HR above the quote and citation on large size.

Placing the rule there is a bit tricky - I'll leave a separate inline review for team discussion

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [inline templates/_macros/quote-wrapper.jinja:131]: @pastelcyborg regarding my [previous comment](https://github.com/canonical/vanilla-framework/pull/5275#issuecomment-2289130361), here's the strangeness around spacing text away from the rule. 

The text of the quote and citation source name are aligned vertically because they all have no top padding. Adding a rule above them as-is makes the rule far too close to the text, so I use vertical spacing utility to compensate. Is this the sort of thing that will require additional scss?

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [inline templates/_macros/quote-wrapper.jinja:120]: Per [previous comment](https://github.com/canonical/vanilla-framework/pull/5275#issuecomment-2289130361), this `u-sv-1` is used to make the top of the image line up with the top of the quote and citation

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja]: I think I had some weird cache weirdness going on when I reported this, as previously `hr` was getting no bottom margin, but now it is spacing the quote and citation out correctly as long as I don't use `p-rule`. So this is tangentially related to #5241 , but is not blocked by it.

**Comment by jmuzina** [comment]
Made an update to add a muted rule between quote and citation on small/medium. 

This will be broken until merge of #5241 , as `p-rule--muted` has no bottom margin currently.

**Comment by jmuzina** [comment]
Using `hr.is-muted` so this isn't blocked by the HR PR - ready for review

**Comment by jmuzina** [comment]
Made an update to add `hr.is-highlighted` (a highlighted rule that has bottom margin) and use it in this pattern.

**Comment by lyubomir-popov** [comment]
I'll continue the review tomorrow, but the top line should not go wider than the fixed width:
image

**Comment by lyubomir-popov** [comment]
- we should not tamper with baseline nudges (by adding u-no-padding--top/u-no-margin--bottom). This misalignes h2 and parsagraph for example. The coment applies across the board, not just this pr:

image

- I need to give you a version of therexroth logo that aligns properly (by inserting white space in the image itself)

- these two should be wrapped by a common parent, with p-section class. Individual quotes in the parent should be spaced using the shallow spacing that is already there

image

- all hrs other than the one above the h2 heading should havethe class "is muted" (we need to add this to the hr docs)
image

- can we please insert these mobile classes to prevent the logo from becoming huge on small screens:
image

**Comment by jmuzina** [comment]
@lyubomir-popov Adjusted

Should the image be col-small-3 perhaps? It gets very small on small now:

Screenshot 2024-08-20 at 9 28 41 AM

**Comment by jmuzina** [comment]
>     * we should not tamper with baseline nudges (by adding u-no-padding--top/u-no-margin--bottom). This misalignes h2 and parsagraph for example. The coment applies across the board, not just this pr:

I got rid of all instances of `u-no-padding--top` and `u-no-margin--bottom` except for the following:

1. `h5.u-no-margin--bottom` on the citation source's name - without this, there is too much space between the citation source name and title to match the design
2. `p.u-no-margin--bottom` on the citation source's title - without this, there is too much space between the source and the content beneath it on small. I could also use `span` (without margin collapse), but that places the text slightly beneath the baseline. Here's how it looks with `p.u-no-margin--bottom`
Screenshot 2024-08-20 at 4 41 01 PM

>     * I need to give you a version of therexroth logo that aligns properly (by inserting white space in the image itself)
Thanks! For the time being I've added you as an assignee while I await that logo.

>     * these two should be wrapped by a common parent, with p-section class. Individual quotes in the parent should be spaced using the shallow spacing that is already there
Done!

>     * all hrs other than the one above the h2 heading should havethe class "is muted" (we need to add this to the hr docs)
Done (changed HR between quote and heading to `hr.is-muted`). 
Also, the [base HR docs](https://vanillaframework.io/docs/base/separators#muted-horizontal-rule) do say this:
"By default, all horizontal rules should be muted. The regular, non-muted variant is only used at the top of content sections, above headings", if that isn't enough let me know and we can address it.

>     * can we please insert these mobile classes to prevent the logo from becoming huge on small screens:
Done, but used `col-small-2` rather than `col-small-1` as discussed at standup today, to prevent the logo from being excessively small.

**Comment by lyubomir-popov** [comment]
@Sophie-32 is there a quote we can use in the vanilla examples? we're currently using rexroth, we probably don't have permisison. Ideally, one of our own with a canonical product logo?

**Comment by jmuzina** [comment]
@lyubomir-popov Requested an H6 variant, adding that shortly - this can be a very trivial addition

**Comment by jmuzina** [comment]
Quote wrapper variant added. I've also added an example showing the minimal content (just the quote text).

**Comment by jmuzina** [comment]
@lyubomir-popov @pastelcyborg @advl I've added notices to the docs pages of all patterns examples stating that patterns are a recent/changing addition:

Screenshot 2024-08-21 at 10 47 45 AM

File: templates/docs/patterns/quote-wrapper/index.md
Comment (jmuzina) [reply templates/docs/patterns/quote-wrapper/index.md:7]: These docs pages are rendered with [mistune](https://mistune.lepture.com/en/latest/guide.html), not Jinja - so we'd need to research & implement some new way to do this. So, it may be a bit out of scope of this PR, but I do agree this is cumbersome

File: templates/docs/patterns/hero/index.md
Comment (jmuzina) [reply templates/docs/patterns/hero/index.md]: Good call, reworded

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja]: Good call, added semicolons there

**Comment by lyubomir-popov** [comment]
I think I found a suitable quote to replace the rexroth content - [here ](https://canonical.com/careers/company-culture/diversity)
image

**Comment by jmuzina** [comment]
@lyubomir-popov I've updated the following:

- Revised the H6 quote to a quote from the 20 years page, provided by @mattea-turic
- Replaced the Rexroth signpost logos with Qualcomm logos (from the 20 years page)
- Replaced the Rexroth full-width cover images with the full-width cover image from our [diversity](https://canonical.com/careers/company-culture/diversity) page

File: templates/docs/patterns/quote-wrapper/index.md
Comment (jmuzina) [reply templates/docs/patterns/quote-wrapper/index.md]: @pastelcyborg I found a way to do this, and I've implemented it. Let me know what you think.
It turns out this can be done with Jinja just like any other template.

**Comment by lyubomir-popov** [comment]
this overrides the h4 margin, which is incorrect - can we remove pls 

image

same on the other text 
image

and we need to use canonical quotes from the page I sent you yesterday, as now the logo, person and role don't match

all of this is also old styling
image

**Comment by lyubomir-popov** [comment]
updated logos to match text baseline at the 3 font-sizes https://assets.ubuntu.com/manager?tag=quote-example-logos

**Comment by jmuzina** [comment]
@lyubomir-popov I've updated the macro to remove the spacing issues that you raised and we discussed in standup, and changed the quote text and citation sources to Canonical personnel from the careers pages.

I've also changed the CTA to "Explore Careers" and the header to "Our hiring process".

Have a look at the [combined example](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/combined?theme=light) and verify every example looks correct, please :)

**Comment by lyubomir-popov** [comment]
no need for u-no-margin--bottom on these 

image

once removed, the baselines of the h4 and second line of the role line up, which is an added nice touch

image

same here:

image

Do we have an example with the header with black line not present when the quote is other than the first?

**Comment by jmuzina** [comment]
> no need for u-no-margin--bottom on these
> 
> image
> 
> once removed, the baselines of the h4 and second line of the role line up, which is an added nice touch
> 
> image
> 
> same here:
> 
> image
> 
> Do we have an example with the header with black line not present when the quote is other than the first?

Just added example of no-header for each of the heading levels

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja:17]: I think all params that accept text input that will be rendered as content should end in `_text` - it's a good way to suggest to the user what the intent is there.

File: templates/docs/examples/patterns/quote-wrapper/h2-no-citation.html
Comment (jmuzina) [reply templates/docs/examples/patterns/quote-wrapper/h2-no-citation.html:15]: I'm of two minds on this sort of convention for slots as they can be much more free-form. 

1. Consider a "description" block that is meant to contain free-form descriptive elements (mostly `<p>`). The information passed into that slot is not necessarily the description's "content", it *is* the description in and of itself. The information there is "content" in the sense of it being "slot content" - but that is more of a concern for the people making the macros than it is those using the macros.
2. In the literal sense of the word, the information passed into the slots is site content. 

I guess I'm just not sure if we can say for certain that every slot should be called `_content`.

Comment (jmuzina) [reply templates/docs/examples/patterns/quote-wrapper/h2-no-citation.html:15]: Here's an example of why I'm hesitant to do this:

Screenshot 2024-08-22 at 3 40 49 PM

This slot is intended for just passing in a link. It feels a bit strange to say `link_content`, as the developer is passing in a link, not the contents of a link.

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja:17]: I've switched all of the quote wrapper's text-only parameters (so, everything but `quote_heading_level`) to end with `_text`. The other patterns should be handled separately, I think - this PR is already quite large. I'll make an issue to cover Hero & Tiered List

**Comment by jmuzina** [comment]
@lyubomir-popov Is hoping to decrease spacing between quote source name and role slightly. Here are two proposed approaches:

Add `u-no-margin--bottom` to both of them, then wrap their row in a shallow section:
Screenshot 2024-08-23 at 9 52 28 AM

OR:

The following change to markup around the name and role:
Screenshot 2024-08-23 at 10 01 12 AM

**Comment by jmuzina** [comment]
@lyubomir-popov I made the change requested in [Figma](https://www.figma.com/design/vH19y2J2YqwUfksebSUvTW?node-id=26-1135&m=dev#923221984) and tried to make it work with the spacing adjustments, have a look and let me know what you think.

`u-no-margin--bottom` was added to the source name to align with what is shown in Figma

File: templates/_macros/quote-wrapper.jinja
Comment (bartaz) [reply templates/_macros/quote-wrapper.jinja:17]: I would also consider abstracting it. Especially that we already seem to limit that just to "2, 4, 6", so it's not like we allow any possible heading level.

It seems to be a quote size option. The fact that it matches to one of heading sizes is a bit of implementation detail.

I guess again, the question is: how is this documented and presented from design perspective - in figma. Is this marked as a heading level in Figma, or as a size of "Quote pattern" in Figma? Because in the end, that's what matters: developers will be looking into Figma, and need to translate this into props of a pattern.

File: templates/docs/examples/patterns/quote-wrapper/h2-no-citation.html
Comment (bartaz) [reply templates/docs/examples/patterns/quote-wrapper/h2-no-citation.html:15]: Just to add to that: if every slot name ends with `_content`, what's the point?

File: templates/_macros/quote-wrapper.jinja
Comment (bartaz) [inline templates/_macros/quote-wrapper.jinja:81]: Why not `row--50-50`?

File: templates/docs/templates/notice.html
Comment (jmuzina) [reply templates/docs/templates/notice.html:1]: Good point, that can be changed to macro

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja:128]: @bartaz  When I use `row--25-75`, the nested grid for the source is broken:

Expected:
Screenshot from 2024-08-27 11-42-11

Observed:
Screenshot from 2024-08-27 11-41-56

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja:139]: See [above comment](https://github.com/canonical/vanilla-framework/pull/5275#discussion_r1733123428)

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja:17]: These were presented as heading levels in Figma. It is not difficult to translate them into a `quote_size` param with options of `large`, `medium`, `small`. However, that would make it difficult for this to support any other heading level sizes later.

File: templates/docs/templates/notice.html
Comment (jmuzina) [reply templates/docs/templates/notice.html]: Changed to use macro

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja:128]: Update: Lyubo's made a [design simplification](https://www.figma.com/design/vH19y2J2YqwUfksebSUvTW?node-id=26-1135&m=dev#923221984) that means this can be done

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja:136]: Switched to 25/75 row

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja]: Switched to 25/75 row

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja]: Abstracted this into `large`/`medium`/`small`, design does not think we will need more sizes ([ref](https://chat.canonical.com/canonical/pl/9qi85kurijr59n9yqu7dni7y5h))

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [inline templates/_macros/quote-wrapper.jinja:132]: Note to self tomorrow: this comment can be removed, as we're now using the 25/75 row, making this obsolete

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [inline templates/_macros/quote-wrapper.jinja:1]: Need to rename this file and namespace the macro with vf_ for compatibility with macro publishing efforts

Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja:1]: Updated

File: templates/_macros/quote-wrapper.jinja
Comment (jmuzina) [reply templates/_macros/quote-wrapper.jinja]: Updated

File: side-navigation.yaml
Comment (bartaz) [inline side-navigation.yaml:120]: If "Tiered list" I guess it should be "Quote wrapper" then
```suggestion
    - title: Quote wrapper
```

File: templates/_macros/hero.jinja
Comment (bartaz) [inline templates/_macros/hero.jinja:75]: Not sure how this change is related to quote?

File: templates/docs/patterns/quote-wrapper/index.md
Comment (bartaz) [inline templates/docs/patterns/quote-wrapper/index.md:25]: Should we document heading sizes here, or macro size values?

**Review by bartaz — COMMENTED** [review]
Overall it looks good.

But in relation to recent discussion on the shape of these patterns (and how flexible they are, etc) I think it would be good to have a spec approved first as well.
Do we have the spec (shape of the parameters, variants, options, etc) written down? It would be good for it to be reviewed and approved by Lyubo, UX, ideally Sophie as well. To avoid a situation when we release this pattern and next thing we get a request to add a layout option or additional image/cta block or whatever.

File: templates/_macros/hero.jinja
Comment (jmuzina) [reply templates/_macros/hero.jinja:75]: This is a drive-by to fix something we missed in design review on hero, as requested by Lyubo [here](https://chat.canonical.com/canonical/pl/1bda8imr5bymbde47unn8hjmoe)

File: templates/docs/patterns/quote-wrapper/index.md
Comment (jmuzina) [reply templates/docs/patterns/quote-wrapper/index.md:25]: In the [PR that shows macro calls in code snippets](https://github.com/canonical/vanilla-framework/pull/5321), we show HTML by default still. So, I figure the best way to do this is to document the HTML classes in the MD sections, and document the Jinja choices using table of params/slots as with #5325

**Comment by jmuzina** [comment]
The team has decided in SU today that we will be starting our HOC speccing efforts by retroactively speccing our to-be-delivered HOCs before they are merged. So, this PR will be blocked until review and approval of a spec by design, UX, and code.

**Comment by jmuzina** [comment]
Made some changes after meeting with @lyubomir-popov to discuss [c.com PR 1380](https://github.com/canonical/canonical.com/pull/1380).

- Design change: Changed the medium layout design: it now uses a 33/66 split with the signpost (or a gap if no signpost is provided) on the left, rest of the content on the right. The header's grid was also adjusted to use 33/66 on medium to line up with the signpost/quote grid.
- Wrapping the quote text and the signpost logo in a shallow section
- Slightly reduced spacing between the header row and rest of the quote
- Fixed some broken grid usage for the signpost logo
- Switched from using base `hr` to `p-rule` component
- Show the HR above the quote text on all breakpoints instead of only on large

File: templates/docs/patterns/quote-wrapper/index.md
Comment (jmuzina) [reply templates/docs/patterns/quote-wrapper/index.md]: Updated to use macro param/slot documentation per established standards from hero and tiered list

**Comment by jmuzina** [comment]
Spec has been approved, putting this PR back into consideration :)

**Comment by bartaz** [comment]
I guess it could be worth updating the pattern to use new 8 col grid? But also, doesn't need to be in scope of this PR. We could address it as follow up (could be done for all patterns I suppose).

**Review by bartaz — APPROVED** [review]
Looks good. We could consider updating this (and other?) patterns to new grid, but this can be a follow up.

**Pull request merged**
